drivers: flash_gecko: Add flash driver for SiLabs Gecko SoCs

mnkp opened the pull request:
This commit adds a flash driver for SiLabs Gecko SoCs.

Tested with SLWSTK6061A / BRD4250B wireless starter kit.

mnkp committed:
drivers: flash_gecko: Add flash driver for SiLabs Gecko SoCs

Tested with SLWSTK6061A / BRD4250B wireless starter kit.

Signed-off-by: Piotr Mienkowski <piotr.mienkowski@gmail.com>

mnkp committed:
boards: efr32_slwstk6061a: add mcuboot flash partitions

This commit adds mcuboot flash partitions for SiLabs
efr32_slwstk6061a evaluation board.

Signed-off-by: Piotr Mienkowski <piotr.mienkowski@gmail.com>

mnkp commented:
MCUboot project has a short piece on [Building and using MCUboot with Zephyr](https://github.com/runtimeco/mcuboot/blob/master/docs/readme-zephyr.md).

mnkp commented:
@chrta if you would be willing to test I could also enable flash driver on EFM32WG SoC in this PR. Unfortunately, as I have mentioned, I don't have this chip myself.

chrta commented:
@mnkp Yes, please do. I can test this. I assume there is some sample code i can run to test flash access.

mnkp commented:
I have added support for EFM32WG SoC and efm32wg_stk3800 board.

A convenient way to test is using samples/drivers/flash_shell/ project. In case of SiLabs flash the block write size is 4 bytes, i.e. the amount of bytes written has to be multiple of 4 or the driver will return an errror. The flash page size / the size of erase block is 2048. Both offset and size parameters of flash_erase have to be multiple of 2048.

chrta commented:
In tested this on the efm32wg_stk3800 board and can confirm that samples/drivers/flash_shell/ works (read, write and erase).

mnkp commented:
@chrta Thanks!

mnkp commented:
I've pushed a new version which adds FLASH to the list of supported features on the EFM32WG-STK3800 and EFR32-SLWSTK6061A board documentation page. This was previously overlooked. No other changes were made.

dbkinder reviewed (approved):
doc update LGTM, thanks.

MaureenHelm commented on drivers/flash/Kconfig:
```diff
@@ -20,20 +24,31 @@ config FLASH_HAS_PAGE_LAYOUT
 
-menuconfig FLASH
-	bool "Flash hardware support"
+config SYS_LOG_FLASH_LEVEL
```
Use the new logging subsystem instead. See #10029

MaureenHelm commented on dts/bindings/flash_controller/silabs,efm32-flash-controller.yaml:
```diff
@@ -0,0 +1,19 @@
+---
```
Add a license/copyright header

galak commented:
Are we able to use this on efm32hg & efr32mg ?

galak commented:
@MaureenHelm can you re-review?

galak reviewed (commented):
Looks good, would like to now if we can use this on the 2 other SiLabs SoCs we now have support for in tree.

ekarlso commented:
> Are we able to use this on efm32hg & efr32mg ?

I think we can add it to after merging this to the HG and MG.

ekarlso commented:
Reference to #9042

mnkp commented:
>> Are we able to use this on efm32hg & efr32mg ?

> I think we can add it to after merging this to the HG and MG.

I could add support for efm32hg & efr32mg now if someone is willing to test it (I don't have necessary hardware).

ekarlso commented:
The board support for 32mg is not added yet, just SoC. I can add it in a seperate PR after this and the board support merges..

mnkp commented:
There are two issues with this PR I somehow overlooked before.

1) MSC_WriteWord and MSC_ErasePage functions are not placed in RAM.

MSC_WriteWord and MSC_ErasePage functions which are provided by SiLabs and located in ext/hal/silabs/gecko/emlib/ folder are supposed to be placed in RAM. That's due to the fact that during flash write/erase operations the flash is inaccessible / MCU cannot read any code from flash.

When compiling I get the warning
`/opt/zephyr-sdk/sysroots/x86_64-pokysdk-linux/usr/libexec/arm-zephyr-eabi/gcc/arm-zephyr-eabi/6.2.0/real-ld: warning: orphan section `.ram' from `zephyr/libzephyr.a(em_msc.c.obj)' being placed in section `.ram'.`

That's because SiLabs code tries to place the two functions in section .ram which is not defined in the Zephyr linker script. The driver sill works as CPU instruction access is simply stalled until the flash becomes available. 

I've looked for some Zephyr macro, etc. to place function in RAM but couldn't find any. Do we have one?

On a related note: a long erase operation could cause SysTick interrupt not to be served and consequently mess up Zephyr's notion of time. I didn't run any extended test however other flash drivers don't seem to bother with this issue either.

2) In flash_shell example project MPU_ALLOW_FLASH_WRITE is set to y even for devices with no MPU support enabled.

While running cmake I get the warning
```
warning: MPU_ALLOW_FLASH_WRITE (defined at arch/arm/core/cortex_m/mpu/Kconfig:26) was assigned the
value 'y' but got the value 'n'. You can check symbol information (including dependencies) in the
'menuconfig' interface [...]
```
This happens because `samples/drivers/flash_shell/prj.conf` is forcing the value of MPU_ALLOW_FLASH_WRITE option to y even if the necessary conditions e.g. ARM_MPU=y are not fulfilled.

Point 2) is not strictly related to this driver but it may well be the first time when the issue is visible and flash_shell example is likely the first project user will try to play with.

mnkp commented:
> The board support for 32mg is not added yet, just SoC.

@ekarlso If you are able to test this PR in your private work area and give an OK I can add support for efm32hg & efr32mg. But we could do it also later.

ekarlso commented:
I can test in the evening @mnkp  :)

mnkp commented:
One more thing. Currently the .yaml file, part of the DTS support, for this driver is called `dts/bindings/flash_controller/silabs,efm32-flash-controller.yaml`. We seem to be using two different naming conventions however. Some silabs .yaml file use gecko in their name e.g. `dts/bindings/serial/silabs,gecko-usart.yaml` others the name of the series e.g. `dts/bindings/gpio/silabs,efm32-gpio.yaml`.

The silabs,gecko-usart.yaml seems to be newer. Should this PR follow this approach? I.e. should I rename the .yaml file to `dts/bindings/flash_controller/silabs,gecko-flash-controller.yaml`?

This is in principle a question if .yaml file should relate to driver they support or the underlying hardware the driver runs on?

ekarlso commented:
@mnkp I have the files already for the mg

chrta commented:
> The silabs,gecko-usart.yaml seems to be newer. Should this PR follow this approach? I.e. should I rename the .yaml file to dts/bindings/flash_controller/silabs,gecko-flash-controller.yaml?

I think using gecko-prefix is better. The supported MCUs are EFM32, EZR32 and EFR32, so the efm32-prefix is to specific imho.

mnkp commented:
I have added support for efm32hg & efr32mg series and renamed the .yaml file to silabs,gecko-flash-controller.yaml.

Note: The convenient way to test the driver is `samples/drivers/flash_shell` project. In case of the efm32hg MCU however the default configuration takes too much RAM to fit into 8kB available on the chip. To successfully compile the project one has to turn off the logger subsystem.

mnkp commented:
Rebased on top of 554001696561fc6f0690db4f01c13b9d18716869 which adds EFR32MG-SLTB004A support.

mnkp commented:
Thanks @MaureenHelm for approving this PR. However, there is still the issue with RAM functions. This driver relies on MSC library module provided by SiLabs which is located in ext/ folder. The module tries to place MSC_WriteWord and MSC_ErasePage functions in .ram section. This section is however missing from Zephyr linker script. That results in a warning about an orphan section and the two functions are placed in flash, not RAM.

During flash write or erase operation the flash is not accessible to the CPU, i.e. if CPU attempts to read an instruction from flash it will stall. If MSC_WriteWord and MSC_ErasePage functions are placed in flash the CPU will stall immediately, if they are placed in RAM we open a possibility to place other critical ISRs in RAM which then can run even if flash is temporarily inaccessible. This is not straightforward as one would need to ensure that no other kernel, ISR code placed in flash is invoked in the meantime but at least we don't close such a possibility. I didn't test it but I expect that due to design of AHB bus a DMA transfer between peripheral and RAM would still be possible even if CPU is stalled.

A page erase operation can take up to 40 ms so it's a pretty long time.

I see following solutions:
1) Modify MSC module code placed in ext/ to use Zephyr's macro to place the function code in RAM.
2) Add .ram section to Zephyr's linker script.
3) Modify MSC module code placed in ext/ not to place the two functions in RAM.
4) Implement the two functions natively in this driver so we don't need to touch ext/ folder neither the Zephyr's linker script.

galak commented:
> 1. MSC_WriteWord and MSC_ErasePage functions are not placed in RAM.

is this issue dealt with, was a little unclear to me.

mnkp commented:
>> MSC_WriteWord and MSC_ErasePage functions are not placed in RAM.

> is this issue dealt with, was a little unclear to me.

No, we still need to fix it. I've detailed possible solutions 1. to 4. above. Other suggestions are welcomed.

What would be a Zephyr way to place a function in RAM? Some macro?

To sum up: the driver works as it is however there are currently two warnings generated when building the driver. One is about an orphan .ram section, the other about inconsistency with the default value of MPU_ALLOW_FLASH_WRITE option (when building flash_shell example project).

mnkp commented:
@ekarlso any luck testing the driver on efm32hg_slstk3400a and efr32mg_sltb004a boards?

ekarlso commented:
@mnkp yes, works as a charm!

ekarlso commented:
@mnkp are we ready to merge this or?

mnkp commented:
@ekarlso Thanks for testing! 

I was going to say that we should fix the orphan .ram section warning but the PR was merged in the meantime.

I will be happy to fix the issue, however I'm not sure how to proceed. I would need some guidance from maintainers. As a reminder here are a few possible choices:

> 1) Modify MSC module code placed in ext/ to use Zephyr's macro to place the function code in RAM.
> 2) Add .ram section to Zephyr's linker script.
> 3) Modify MSC module code placed in ext/ not to place the two functions in RAM.
> 4) Implement the two functions natively in this driver so we don't need to touch ext/ folder neither the Zephyr's linker script.

Do we have a method in Zephyr to force placing a function in RAM?
